pmd-lang-test/src/main/kotlin/net/sourceforge/pmd/cpd/test/CpdTextComparisonTest.kt

```diff
@@ -0,0 +1,146 @@
+
+    /**
+     * A test comparing the output of the tokenizer.
+     *
+     * @param fileBaseName   Name of the source file (without extension or resource prefix)
+     * @param expectedSuffix Suffix to append to the expected file. This allows reusing the same source file
+     *                       with different configurations, provided the suffix is different
+     * @param properties     Properties to configure [newTokenizer]
+     */
+    @JvmOverloads
+    fun doTest(fileBaseName: String, expectedSuffix: String = "", properties: Properties = defaultProperties()) {
+        super.doTest(fileBaseName, expectedSuffix) { sourceText ->
+            val sourceCode = SourceCode(SourceCode.StringCodeLoader(sourceText, "$fileBaseName$extensionIncludingDot"))
+            val tokens = Tokens().also {
+                val tokenizer = newTokenizer(properties)
+                tokenizer.tokenize(sourceCode, it)
+            }
+
+            buildString { format(tokens) }
+        }
+    }
+
+
+    private fun StringBuilder.format(tokens: Tokens) {
+        appendHeader().appendln()
+
+        var curLine = -1
+
+        for (token in tokens.iterator()) {
+
+            if (token === TokenEntry.EOF) {
+                append("EOF").appendln()
+                continue
+            }
+
+            if (curLine != token.beginLine) {
+                curLine = token.beginLine
+                append('L').append(curLine).appendln()
+            }
+
+            formatLine(token).appendln()
+        }
+    }
+
+
+    private fun StringBuilder.appendHeader() =
+            formatLine(
+                    escapedImage = "[Image] or [Truncated image[",
+                    bcol = "Bcol",
+                    ecol = "Ecol"
+            )
+
+
+    private fun StringBuilder.formatLine(token: TokenEntry) =
+            formatLine(
+                    escapedImage = escapeImage(token.toString()),
+                    bcol = token.beginColumn,
+                    ecol = token.endColumn
+            )
+
+
+    private fun StringBuilder.formatLine(escapedImage: String, bcol: Any, ecol: Any): StringBuilder {
+        var colStart = length
+        colStart = append(Indent).append(escapedImage).padCol(colStart, Col0Width)
+        colStart = append(Indent).append(bcol).padCol(colStart, Col1Width)
+        return append(ecol)
+    }
+
+    private fun StringBuilder.padCol(colStart: Int, colWidth: Int): Int {
+        for (i in 1..(colStart + colWidth - this.length))
+            append(' ')
+
+        return length
+    }
+
+
+    private fun escapeImage(str: String): String {
+        val escaped = str
+                .replace("\\", "\\\\")                 // escape backslashes
+                .replace("\r\n", "\\r\\n")             // CRLF (treated specially because it has a different length)
+                .replace("\t", "\\t")                  // TAB
+                .replace(Regex("\\R"), "\\\\n")       // escape other newlines (normalizing)
```

Comment:
Does `\\R` work here? It maybe just doesn't match anything (and we don't have anything else than `\n` anyway...).
I'll change it to `\\u000D\\u000A|[\\u000A\\u000B\\u000C\\u000D\\u0085\\u2028\\u2029]` with the comment "java8+ ..."